walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
+        HashSet<String> allAttendees = new HashSet<>(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        Collection<TimeRange> withOptionalAttendees = getMeetingSlots(events, allAttendees, request.getDuration());
+        
+        // if there are slots on which both optional and mandatory attendees are free - return them
+        // if not - return only slots on which mandatory attendees are free
+        if (withOptionalAttendees.isEmpty()) {
+            return getMeetingSlots(events, request.getAttendees(), request.getDuration());
+        } else {
+            return withOptionalAttendees;
+        }
+    }
+
+    /**
+     * Function to get all possible slots for a meeting of specified group of people,
+     * with no optional attendees.
+     */
+    Collection<TimeRange> getMeetingSlots(Collection<Event> events, Collection<String> attendees, long duration) {
+        // make an array for getEmptyTimeRanges function
+        // put there all significant events' START ad END points assuming that event is a segment on time line.
+        ArrayList<Point> meetingsPoints = new ArrayList<>();
+
+        for (Event event : events) {
+            if (doIntersect(event.getAttendees(), attendees)) {
+                meetingsPoints.add(new Point(Point.Type.START, event));
+                meetingsPoints.add(new Point(Point.Type.END, event));
+            }
+        }
+        
+        return getEmptyTimeRanges(meetingsPoints, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, duration);
+    }
+
+    /**
+     * @return all segments from firstPoint to lastPoint inclusive which are no shorter than minLength
+     *.        and do not intersect with segments from intersectingSegments
+     * @param intersectingSegments - array of Points. To put segment in this array, you should put its
+     *.                              START and END Points there.
+     * @param firstPoint - beginning of all time. Returned collection will not contain time before it.
+     * @param lastPoint - end of all time. Returned collection will not contain time after it.
+     * @param minLength - minimal length for each of the returned intervals
+     * Algorithm:
+     *.        All empty slots can start only where some segment ends and end only where some segment starts (or at
+               firstPoint or lastPoint)
+     *.        There cannot be any other points from intersectingSegments inside the empty slot - only its start end end.
+     *         Sort all points in intersectingSegments in ascending order and look at them one by one.
+     *.        Count how many segments are currently covering the point you are looking at: get a counter
+     *.        which increments every time you look at START point and decrements every time you look at END point.
+     *.        If the point is covered with 0 segments - it is a start of segment which can be one of the answer
```

Comment:
Don't you mean "it is an END of segment which can be one of the answer"? It is a start of an intersectingSegment, but an end of one which can be the answer.

Reply:
Actually, it can be both an END and a START. If it is covered with 0 segments, we only know that it belongs to empty interval. Changed it :)